metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,503 @@
+              otherQual
+            )
+              +: Enumerator.Guard.apply(
+                if (isFilter) nextCondition
+                else
+                  Term.ApplyUnary.apply(Term.Name.apply("!"), nextCondition)
+              ) +: elems,
+            maybeYieldTerm
+          )
+      }
+    }
+    result.getOrElse(List.empty, currentYieldTerm)
+  }
+
+  private def extractChainedForYield(
+      perhapsLastName: Option[String],
+      currentYieldTerm: Option[Term],
+      existingForElements: List[Enumerator],
+      termApply: Term.Apply,
+      nameGenerator: MetalsNames
+  ): (List[Enumerator], Option[Term]) = {
+    val perhapsValueNameAndNextQual =
+      processValueNameAndNextQual(
+        termApply.args.head,
+        nameGenerator
+      )
+
+    termApply.fun match {
+      case termSelect: Term.Select
+          if termSelect.name.value == "flatMap" || termSelect.name.value == "map" =>
+        val shouldFlat = termSelect.name.value == "flatMap"
+
+        val result = for {
+          valueName <- perhapsValueNameAndNextQual.map(_._1)
+          nextQual <- perhapsValueNameAndNextQual.map(_._2)
+        } yield {
+          val (elems, maybeYieldTerm) =
+            obtainNextYieldAndElemsForMap(
+              nameGenerator,
+              perhapsLastName,
+              shouldFlat,
+              existingForElements,
+              currentYieldTerm,
+              nextQual
+            )
+
+          processMap(
+            elems,
+            maybeYieldTerm,
+            nameGenerator,
+            valueName,
+            termSelect.qual
+          )
+        }
+        result.getOrElse(List.empty, None)
+
+      case termSelect: Term.Select
+          if termSelect.name.value == "filter" || termSelect.name.value == "filterNot" ||
+            termSelect.name.value == "withFilter" =>
+        val isFilter =
+          termSelect.name.value == "filter" || termSelect.name.value == "withFilter"
+        processFilter(
+          perhapsValueNameAndNextQual,
+          nameGenerator,
+          perhapsLastName,
+          isFilter,
+          existingForElements,
+          currentYieldTerm,
+          termSelect.qual
+        )
+
+      case _ =>
+        perhapsLastName match {
+          case Some(lastName) =>
+            (
+              Enumerator.Generator(
+                Pat.Var.apply(Term.Name.apply(lastName)),
+                termApply
+              )
+                +: existingForElements,
+              currentYieldTerm
+            )
+          case None =>
+            (existingForElements, Some(termApply))
+        }
+    }
+  }
+
+  private def getIndentForPos(
+      treePos: Position,
+      document: String
+  ): String =
+    document
+      .substring(treePos.start - treePos.startColumn, treePos.start)
+      .takeWhile(_.isWhitespace)
+
+}
+
+object FlatMapToForComprehensionCodeAction {
+  val flatMapToForComprehension = "Turn into for comprehension"
```

Comment:
```suggestion
  val flatMapToForComprehension = "Convert to for comprehension"
```

what do you think the name ?

Reply:
Perfect 💯